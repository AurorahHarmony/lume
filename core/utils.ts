import { DOMParser, HTMLDocument } from "../deps/dom.ts";
import { brightGreen, cyan, dim, gray, green, red } from "../deps/colors.ts";
import { dirname, extname, join, posix, SEP } from "../deps/path.ts";
import { parse } from "../deps/jsonc.ts";

/** A list of the available optional plugins */
export const pluginNames = [
  "attributes",
  "base_path",
  "code_highlight",
  "date",
  "esbuild",
  "eta",
  "feed",
  "filter_pages",
  "imagick",
  "inline",
  "jsx",
  "jsx_preact",
  "katex",
  "lightningcss",
  "liquid",
  "mdx",
  "metas",
  "minify_html",
  "modify_urls",
  "multilanguage",
  "nav",
  "decap_cms",
  "on_demand",
  "pagefind",
  "picture",
  "postcss",
  "prism",
  "pug",
  "relations",
  "relative_urls",
  "remark",
  "resolve_urls",
  "sass",
  "sheets",
  "sitemap",
  "slugify_urls",
  "source_maps",
  "svgo",
  "tailwindcss",
  "terser",
  "toml",
  "unocss",
  "vento",
];

export function log(...lines: (string | undefined)[]) {
  console.log("----------------------------------------");
  lines.forEach((line) => line && console.log(line));
  console.log("----------------------------------------");
}

/** Check the compatibility with the current Deno version */
export function checkDenoVersion(): void {
  const minimum = "1.33.4";
  const current = Deno.version.deno;

  if (current < minimum) {
    console.log("----------------------------------------");
    console.error(red("Your Deno version is not compatible with Lume"));
    console.log(`Lume needs Deno ${green(minimum)} or greater`);
    console.log(`Your current version is ${red(current)}`);
    console.log(`Run ${cyan("deno upgrade")} and try again`);
    console.log("----------------------------------------");
    Deno.exit(1);
  }
}

/** Check if the current version is outdated */
export async function checkUpgrade(): Promise<void> {
  const current = getLumeVersion();

  // It's a local version
  if (current.startsWith("local ")) {
    return;
  }

  const stable = !!current.match(/^v\d+\./);
  const expires = 1000 * 60 * 60 * 24; // 1 day
  const interval = localStorage.getItem("lume-upgrade");

  if (interval && parseInt(interval) + expires > Date.now()) {
    return;
  }

  localStorage.setItem("lume-upgrade", Date.now().toString());

  const latest = stable
    ? await getLatestVersion()
    : await getLatestDevelopmentVersion();

  if (current === latest) {
    return;
  }

  const command = "deno task lume upgrade" + (stable ? "" : " --dev");

  log(
    `Update available ${dim(current)} â†’ ${green(latest)}`,
    `Run ${cyan(command)} to update`,
  );
}

/** Return the latest stable version from the deno.land/x repository */
export async function getLatestVersion(): Promise<string> {
  const response = await fetch("https://cdn.deno.land/lume/meta/versions.json");
  const versions = await response.json();
  return versions.latest;
}

/** Return the hash of the latest commit from the GitHub repository */
export async function getLatestDevelopmentVersion(
  branch = "master",
): Promise<string> {
  const response = await fetch(
    `https://api.github.com/repos/lumeland/lume/commits/${branch}`,
  );
  const commits = await response.json();
  return commits.sha;
}

/** Import map file */
export interface ImportMap {
  imports: Record<string, string>;
  scopes?: Record<string, Record<string, string>>;
}

/** Basic options for deno.json file */
export interface DenoConfig {
  importMap?: string;
  tasks?: Record<string, string>;
  compilerOptions?: {
    jsx?: "jsx" | "react-jsx";
    jsxImportSource?: string;
  };
  [key: string]: unknown;
}

/** Run a callback concurrently with all the elements of an Iterable */
export async function concurrent<Type>(
  iterable: AsyncIterable<Type> | Iterable<Type>,
  iteratorFn: (arg: Type) => Promise<unknown>,
  limit = 200,
) {
  const executing: Promise<unknown>[] = [];

  for await (const item of iterable) {
    const p: Promise<unknown> = iteratorFn(item).then(() =>
      executing.splice(executing.indexOf(p), 1)
    );

    executing.push(p);

    if (executing.length >= limit) {
      await Promise.race(executing);
    }
  }

  await Promise.all(executing);
}

const decoder = new TextDecoder();
const encoder = new TextEncoder();

/** Encode a message using SHA-1 algorithm */
export async function sha1(message: string | Uint8Array): Promise<string> {
  if (typeof message === "string") {
    message = encoder.encode(message);
  }

  const hash = await crypto.subtle.digest("SHA-1", message);
  return decoder.decode(hash);
}

/** Helper to create optional properties recursively */
export type DeepPartial<T> = T extends object ? {
    [P in keyof T]?: DeepPartial<T[P]>;
  }
  : T;

/**
 * Merge two objects recursively.
 * It's used to merge user options with default options.
 */
export function merge<Type>(
  defaults: Type,
  user?: Partial<Type> | DeepPartial<Type>,
): Type {
  const merged = { ...defaults };

  if (!user) {
    return merged;
  }

  for (const [key, value] of Object.entries(user)) {
    if (value === undefined) {
      continue;
    }

    // @ts-ignore: No index signature with a parameter of type 'string' was found on type 'unknown'
    if (isPlainObject(merged[key]) && isPlainObject(value)) {
      // @ts-ignore: Type 'string' cannot be used to index type 'Type'
      merged[key] = merge(merged[key], value);
      continue;
    }

    // @ts-ignore: Type 'string' cannot be used to index type 'Type'
    merged[key] = value;
  }

  return merged;
}

const reactElement = Symbol.for("react.element");
const objectConstructor = {}.constructor;

/** Check if the argument passed is a plain object */
export function isPlainObject(obj: unknown): obj is Record<string, unknown> {
  return typeof obj === "object" && obj !== null &&
    obj.constructor === objectConstructor &&
    // @ts-ignore: Check if the argument passed is a React element
    obj["$$typeof"] !== reactElement &&
    // @ts-ignore: Check if the argument passed is a Page.data object
    obj !== obj.page?.data;
}

/**
 * Convert the Windows paths (that use the separator "\")
 * to Posix paths (with the separator "/")
 * and ensure it starts with "/".
 */
export function normalizePath(path: string, rootToRemove?: string) {
  if (rootToRemove) {
    path = path.replace(rootToRemove, "");
  }

  if (SEP !== "/") {
    path = path.replaceAll(SEP, "/");

    // Is absolute Windows path (C:/...)
    if (path.includes(":/")) {
      if (rootToRemove && path.startsWith(rootToRemove)) {
        return posix.join("/", path.replace(rootToRemove, ""));
      }

      return path;
    }
  }

  const absolute = posix.join("/", path);
  return rootToRemove && absolute.startsWith(rootToRemove)
    ? posix.join("/", absolute.replace(rootToRemove, ""))
    : absolute;
}

/** Convert an HTMLDocument instance to a string */
export function documentToString(document: HTMLDocument) {
  const { doctype, documentElement } = document;

  if (!doctype) {
    return `<!DOCTYPE html>\n${documentElement?.outerHTML || ""}`;
  }

  return `<!DOCTYPE ${doctype.name}` +
    (doctype.publicId ? ` PUBLIC "${doctype.publicId}"` : "") +
    (!doctype.publicId && doctype.systemId ? " SYSTEM" : "") +
    (doctype.systemId ? ` "${doctype.systemId}"` : "") +
    `>\n${documentElement?.outerHTML}`;
}

const parser = new DOMParser();

/** Parse a string with HTML code and return an HTMLDocument */
export function stringToDocument(string: string): HTMLDocument {
  const document = parser.parseFromString(string, "text/html");

  if (!document) {
    throw new Error("Unable to parse the HTML code");
  }

  return document;
}

/** Return the current installed version */
export function getLumeVersion(
  url = new URL(import.meta.resolve("../")),
): string {
  const { pathname } = url;
  return pathname.match(/@([^/]+)/)?.[1] ?? `local (${pathname})`;
}

/** Returns the _config file of a site */
export async function getConfigFile(
  path?: string,
): Promise<string | undefined> {
  if (path) {
    try {
      return await Deno.realPath(path);
    } catch {
      throw new Error(`Config file not found (${path})`);
    }
  }

  const paths = ["_config.js", "_config.ts"];

  for (const path of paths) {
    try {
      return await Deno.realPath(path);
    } catch {
      // Ignore
    }
  }
}

export interface DenoConfigResult {
  file: string;
  config: DenoConfig;
  importMap?: ImportMap;
}

/** Detect and returns the Deno configuration */
export async function readDenoConfig(): Promise<DenoConfigResult | undefined> {
  for (const file of ["deno.json", "deno.jsonc"]) {
    try {
      const content = await Deno.readTextFile(file);
      const config = parse(content) as DenoConfig;
      let importMap: ImportMap | undefined;

      if (config.importMap) {
        importMap = isUrl(config.importMap)
          ? await (await fetch(config.importMap)).json()
          : await JSON.parse(await Deno.readTextFile(config.importMap));
      } else if (config.imports) {
        importMap = {
          imports: config.imports as Record<string, string>,
          scopes: config.scopes as Record<string, Record<string, string>>,
        };
      }
      return { file, config, importMap };
    } catch (err) {
      if (err instanceof Deno.errors.NotFound) {
        continue;
      }

      throw err;
    }
  }
}

export function updateLumeVersion(url: URL, denoConfig: DenoConfigResult) {
  denoConfig.importMap ??= { imports: {} };

  const { config, importMap } = denoConfig;

  // Configure the import map
  if (Deno.version.deno < "1.30.0") {
    config.importMap ||= "./import_map.json";
  }

  const oldUrl = importMap.imports["lume/"];
  const newUrl = new URL("./", url).href;
  importMap.imports["lume/"] = newUrl;

  for (const [specifier, url] of Object.entries(importMap.imports)) {
    if (url.startsWith(oldUrl)) {
      importMap.imports[specifier] = url.replace(oldUrl, newUrl);
    }
  }

  // Configure lume tasks
  const tasks = config.tasks || {};
  if (!tasks.lume || !tasks.lume.includes(`echo "import 'lume/cli.ts'"`)) {
    tasks.lume = `echo "import 'lume/cli.ts'" | deno run --unstable -A -`;
    tasks.build = "deno task lume";
    tasks.serve = "deno task lume -s";
  }
  config.tasks = tasks;
}

/** Update the Deno configuration */
export async function writeDenoConfig(options: DenoConfigResult) {
  const { file, config, importMap } = options;

  if (importMap && !config.importMap) {
    config.imports = importMap.imports;
    config.scopes = importMap.scopes;
  }

  if (config.importMap) {
    const importMapFile = join(dirname(file), config.importMap);
    await Deno.writeTextFile(
      importMapFile,
      JSON.stringify(importMap, null, 2) + "\n",
    );
    console.log(brightGreen("Import map file saved:"), importMapFile);
  }

  if (extname(file) === ".jsonc") {
    const save = confirm(
      "Saving the deno.jsonc file will overwrite the comments. Continue?",
    );

    if (!save) {
      console.log(
        "You have to update your deno.jsonc file manually with the following content:",
      );
      console.log(dim(JSON.stringify(config, null, 2)));
      console.log("Use deno.json to update it automatically without asking.");
      return;
    }
  }
  await Deno.writeTextFile(file, JSON.stringify(config, null, 2) + "\n");
  console.log("Deno configuration file saved:", gray(file));
}

export function isUrl(path: string): boolean {
  return !!path.match(/^(https?|file):\/\//);
}
export function isAbsolutePath(path: string): boolean {
  return SEP !== "/" ? /^\w:[\/\\]/.test(path) : path.startsWith("/");
}

export function replaceExtension(
  path: string | false,
  ext: string,
): string | false {
  if (!path) {
    return false;
  }
  return path.replace(/\.\w+$/, ext);
}

export function getPathAndExtension(path: string): [string, string] {
  const extension = getExtension(path);
  const pathWithoutExtension = path.slice(0, -extension.length);
  return [pathWithoutExtension, extension];
}

export function getExtension(path: string): string {
  const match = path.match(/\.\w+$/);
  return match ? match[0] : "";
}

export async function read(
  path: string,
  isBinary: boolean,
): Promise<Uint8Array | string>;
export async function read(
  path: string,
  isBinary: true,
  init?: RequestInit,
): Promise<Uint8Array>;
export async function read(
  path: string,
  isBinary: false,
  init?: RequestInit,
): Promise<string>;
export async function read(
  path: string,
  isBinary: boolean,
  init?: RequestInit,
): Promise<string | Uint8Array> {
  if (!isUrl(path)) {
    if (path.startsWith("data:")) {
      const response = await fetch(path);

      return isBinary
        ? new Uint8Array(await response.arrayBuffer())
        : response.text();
    }

    return isBinary ? Deno.readFile(path) : Deno.readTextFile(path);
  }

  const url = new URL(path);

  if (url.protocol === "file:") {
    return isBinary ? Deno.readFile(url) : Deno.readTextFile(url);
  }

  const cache = await caches.open("lume_remote_files");

  // Prevent https://github.com/denoland/deno/issues/19696
  try {
    const cached = await cache.match(url);

    if (cached) {
      return isBinary
        ? new Uint8Array(await cached.arrayBuffer())
        : cached.text();
    }
  } catch {
    // ignore
  }

  const response = await fetch(url, init);
  await cache.put(url, response.clone());

  return isBinary
    ? new Uint8Array(await response.arrayBuffer())
    : response.text();
}

/**
 * Check if the content variable is a generator.
 */
export function isGenerator(
  content: unknown,
): content is GeneratorFunction | AsyncGeneratorFunction {
  if (typeof content !== "function") {
    return false;
  }

  const name = content.constructor.name;
  return (name === "GeneratorFunction" || name === "AsyncGeneratorFunction");
}

/**
 * Resolve the path of an included file
 * Used in the template engines and processors
 */
export function resolveInclude(
  path: string,
  includesDir: string,
  fromDir?: string,
  rootToRemove?: string,
): string {
  if (isUrl(path)) {
    return path;
  }

  if (path.startsWith(".")) {
    if (!fromDir) {
      throw new Error(`Cannot load "${path}" without a base directory`);
    }

    return normalizePath(posix.join(fromDir, path), rootToRemove);
  }

  const normalized = normalizePath(path, rootToRemove);

  return normalized.startsWith(normalizePath(posix.join(includesDir, "/")))
    ? normalized
    : normalizePath(posix.join(includesDir, normalized));
}

export function env<T>(name: string): T | undefined {
  const value = Deno.env.get(name);

  if (typeof value === "undefined") {
    return undefined;
  }

  switch (value.toLowerCase()) {
    case "true":
    case "on":
    case "1":
      return true as T;

    case "false":
    case "off":
    case "0":
      return false as T;

    default:
      return value as T;
  }
}

export function subExtensions(extensions: string[], subextension?: string) {
  if (!subextension) {
    return extensions;
  }
  if (!subextension.startsWith(".")) {
    throw new Error(
      `Invalid subextension: "${subextension}". It must start with "."`,
    );
  }
  return extensions.map((ext) => subextension + ext);
}
