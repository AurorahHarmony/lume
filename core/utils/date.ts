import { Temporal } from "../../deps/temporal.ts";

/**
 * Returns the date of the git commit that created or modified the file.
 * Thanks to https://github.com/11ty/eleventy/blob/8dd2a1012de92c5ee1eab7c37e6bf1b36183927e/src/Util/DateGitLastUpdated.js
 */
export function getGitDate(
  type: "created" | "modified",
  file: string,
): Date | undefined {
  const args = type === "created"
    ? ["log", "--diff-filter=A", "--follow", "-1", "--format=%at", "--", file]
    : ["log", "-1", "--format=%at", "--", file];

  const { stdout, success } = new Deno.Command("git", { args }).outputSync();

  if (!success) {
    return;
  }
  const str = new TextDecoder().decode(stdout);
  const timestamp = parseInt(str) * 1000;

  if (timestamp) {
    return new Date(timestamp);
  }
}

/** Parse a string to Date */
export function parseDate(date: string): Date {
  return new Date(getZonedDateTime(date).epochMilliseconds);
}

/** Parse a string to a zoned datetime */
export function getZonedDateTime(
  date: string,
  timezone = "UTC",
): Temporal.ZonedDateTime {
  try {
    return Temporal.Instant.from(date).toZonedDateTimeISO(timezone);
  } catch {
    return Temporal.PlainDateTime.from(date).toZonedDateTime(timezone);
  }
}
