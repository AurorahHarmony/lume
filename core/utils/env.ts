export function env<T>(name: string): T | undefined {
  const value = Deno.env.get(name);

  if (typeof value === "undefined") {
    return undefined;
  }

  switch (value.toLowerCase()) {
    case "true":
    case "on":
    case "1":
      return true as T;

    case "false":
    case "off":
    case "0":
      return false as T;

    default:
      return value as T;
  }
}
