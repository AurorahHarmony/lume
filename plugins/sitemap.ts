import { merge } from "../core/utils/object.ts";
import { Page } from "../core/file.ts";
import { stringify } from "../deps/xml.ts";

import type Site from "../core/site.ts";
import type { Data, StaticFile } from "../core/file.ts";

type ChangeFreq =
  | "always"
  | "hourly"
  | "daily"
  | "weekly"
  | "monthly"
  | "yearly"
  | "never";

export interface Options {
  /** The sitemap file name */
  filename?: string;

  /** The query to search pages included in the sitemap */
  query?: string;

  /** The values to sort the sitemap */
  sort?: string;

  /** The key to use for the lastmod field or a custom function */
  lastmod?: string | ((data: Data) => Date);

  /** The key to use for the changefreq field or a custom function */
  changefreq?: string | ((data: Data) => ChangeFreq);

  /** The key to use for the priority field or a custom function */
  priority?: string | ((data: Data) => number);
}

// Default options
export const defaults: Options = {
  filename: "/sitemap.xml",
  query: "",
  sort: "url=asc",
  lastmod: "date",
};

/** A plugin to generate a sitemap.xml from page files after build */
export default function (userOptions?: Options) {
  const options = merge(defaults, userOptions);

  return (site: Site) => {
    site.addEventListener("afterRender", () => {
      // Create the sitemap.xml page
      const sitemap = Page.create(
        options.filename,
        {
          content: generateSitemap(
            site.searcher.pages(options.query, options.sort),
          ),
        },
      );

      // Add to the sitemap page to pages
      site.pages.push(sitemap);

      // Search for the `robots.txt` file
      const robots = site.files.some((file: StaticFile) =>
        file.outputPath === "/robots.txt"
      );

      // If the `robots.txt` file doesn't exist, create it
      if (!robots) {
        const robots = site.pages.find((page: Page) =>
          page.data.url === "/robots.txt"
        );

        if (robots) {
          robots.content += `Sitemap: ${site.url(options.filename, true)}`;
        } else {
          site.pages.push(Page.create(
            "/robots.txt",
            {
              content: `User-agent: *\nAllow: /\n\nSitemap: ${
                site.url("/sitemap.xml", true)
              }`,
            },
          ));
        }
      }
    });

    function generateSitemap(pages: Data[]): string {
      const sitemap = {
        xml: {
          "@version": "1.0",
          "@encoding": "UTF-8",
        },
        urlset: {
          "@xmlns": "http://www.sitemaps.org/schemas/sitemap/0.9",
          url: pages.map((data) => {
            const node: UrlItem = {
              loc: site.url(data.url as string, true),
            };

            const lastmod = getValue<Date>(data, options.lastmod)
              ?.toISOString();
            if (lastmod) {
              node.lastmod = lastmod;
            }

            const changefreq = getValue<ChangeFreq>(data, options.changefreq);
            if (changefreq) {
              node.changefreq = changefreq;
            }

            const priority = getValue<number>(data, options.priority);
            if (priority) {
              node.priority = priority;
            }

            if (data.alternates?.length) {
              node["xhtml:link"] = data.alternates.map((alternate: Data) => ({
                "@rel": "alternate",
                "@hreflang": alternate.lang,
                "@href": site.url(alternate.url as string, true),
              }));
            }
            return node;
          }),
        },
      };

      return stringify(sitemap);
    }
  };
}

interface UrlItem {
  loc: string;
  lastmod?: string;
  changefreq?: ChangeFreq;
  priority?: number;
  "xhtml:link"?: {
    "@rel": "alternate";
    "@hreflang": string;
    "@href": string;
  }[];
}

function getValue<T>(
  data: Data,
  key?: string | ((data: Data) => T),
): T | undefined {
  if (!key) {
    return undefined;
  }

  if (typeof key === "function") {
    return key(data);
  }

  return data[key];
}
