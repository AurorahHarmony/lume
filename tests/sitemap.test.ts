import { assertSiteSnapshot, build, getSite } from "./utils.ts";
import sitemap from "../plugins/sitemap.ts";

Deno.test("Sitemap plugin", async (t) => {
  const site = getSite({
    src: "normal",
    location: new URL("https://example.com/"),
    server: {
      page404: "/page5/",
    },
  });

  site.use(sitemap());
  site.ignore("static.yml");

  await build(site);
  await assertSiteSnapshot(t, site);
});
