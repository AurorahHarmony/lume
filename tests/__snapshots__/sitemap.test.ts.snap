export const snapshot = {};

snapshot[`Sitemap plugin 1`] = `
{
  formats: [
    {
      engines: 1,
      ext: ".tmpl.ts",
      pageLoader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".tmpl.js",
      pageLoader: [AsyncFunction: module],
    },
    {
      engines: undefined,
      ext: ".tmpl.jsonc",
      pageLoader: [AsyncFunction: json],
    },
    {
      engines: undefined,
      ext: ".tmpl.json",
      pageLoader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: undefined,
      ext: ".json",
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: undefined,
      ext: ".jsonc",
    },
    {
      engines: 1,
      ext: ".md",
      pageLoader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".markdown",
      pageLoader: [AsyncFunction: text],
    },
    {
      componentLoader: [AsyncFunction: module],
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
    },
    {
      componentLoader: [AsyncFunction: module],
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
    },
    {
      componentLoader: [AsyncFunction: text],
      engines: 1,
      ext: ".njk",
      includesPath: "_includes",
      pageLoader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: toml],
      engines: undefined,
      ext: ".toml",
      pageLoader: [AsyncFunction: toml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: undefined,
      ext: ".yaml",
      pageLoader: [AsyncFunction: yaml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: undefined,
      ext: ".yml",
      pageLoader: [AsyncFunction: yaml],
    },
  ],
  src: [
    "/",
    "/_data.yml",
    "/favicon.png",
    "/page5.yaml",
    "/pages",
    "/pages/2020-06-21_page2.tmpl.json",
    "/pages/2021-01-02-18-32_page4.tmpl.ts",
    "/pages/2021-12-29-page6.md",
    "/pages/_data",
    "/pages/_data.yml",
    "/pages/_data/colors.yml",
    "/pages/_data/documents.ts",
    "/pages/_data/drinks.js",
    "/pages/_data/names.json",
    "/pages/page1.md",
    "/pages/page3.tmpl.js",
    "/pages/subpage",
    "/pages/subpage/_data.yml",
    "/pages/subpage/page7.tmpl.js",
    "/static.yml",
    "/styles.css",
  ],
}
`;

snapshot[`Sitemap plugin 2`] = `[]`;

snapshot[`Sitemap plugin 3`] = `
[
  {
    content: "User-agent: *
Allow: /

Sitemap: https://example.com/sitemap.xml",
    data: {
      content: "User-agent: *
Allow: /

Sitemap: https://example.com/sitemap.xml",
      page: [
        "src",
        "data",
      ],
      url: "/robots.txt",
    },
    src: {
      asset: true,
      ext: undefined,
      path: "",
      remote: undefined,
      slug: "robots",
    },
  },
  {
    content: '<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
  <url>
    <loc>https://example.com/overrided-page2/</loc>
    <lastmod>2020-06-21T00:00:00.000Z</lastmod>
  </url>
  <url>
    <loc>https://example.com/pages/page4/</loc>
    <lastmod>2021-01-02T18:32:00.000Z</lastmod>
  </url>
  <url>
    <loc>https://example.com/pages/page6/</loc>
    <lastmod>2022-01-01T00:00:00.000Z</lastmod>
  </url>
  <url>
    <loc>https://example.com/pages/subpage/page7/</loc>
    <lastmod>2022-01-02T00:00:00.000Z</lastmod>
  </url>
</urlset>',
    data: {
      content: '<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
  <url>
    <loc>https://example.com/overrided-page2/</loc>
    <lastmod>2020-06-21T00:00:00.000Z</lastmod>
  </url>
  <url>
    <loc>https://example.com/pages/page4/</loc>
    <lastmod>2021-01-02T18:32:00.000Z</lastmod>
  </url>
  <url>
    <loc>https://example.com/pages/page6/</loc>
    <lastmod>2022-01-01T00:00:00.000Z</lastmod>
  </url>
  <url>
    <loc>https://example.com/pages/subpage/page7/</loc>
    <lastmod>2022-01-02T00:00:00.000Z</lastmod>
  </url>
</urlset>',
      page: [
        "src",
        "data",
      ],
      url: "/sitemap.xml",
    },
    src: {
      asset: true,
      ext: undefined,
      path: "",
      remote: undefined,
      slug: "sitemap",
    },
  },
  {
    content: "<!DOCTYPE html>
Content of Page 5",
    data: {
      children: "Content of Page 5",
      content: "Content of Page 5",
      date: [],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      site: "Default site name",
      tags: "Array(0)",
      title: "Page 5",
      url: "/page5/",
    },
    src: {
      asset: undefined,
      ext: ".yaml",
      path: "/page5",
      remote: undefined,
      slug: "page5",
    },
  },
  {
    content: "<!DOCTYPE html>
Content of Page 2",
    data: {
      children: "Content of Page 2",
      colors: "Array(3)",
      content: "Content of Page 2",
      date: [],
      documents: "Array(3)",
      drinks: [
        "alcoholic",
        "others",
      ],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      names: "Array(2)",
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      site: "Folder overrided site name",
      tags: "Array(1)",
      title: "Page 2",
      url: "/overrided-page2/",
    },
    src: {
      asset: undefined,
      ext: ".tmpl.json",
      path: "/pages/2020-06-21_page2",
      remote: undefined,
      slug: "page2",
    },
  },
  {
    content: "<!DOCTYPE html>
Content of Page 4 in Overrided site name, from the file /pages/2021-01-02-18-32_page4.tmpl.ts",
    data: {
      children: "Content of Page 4 in Overrided site name, from the file /pages/2021-01-02-18-32_page4.tmpl.ts",
      colors: "Array(3)",
      content: "default",
      date: [],
      documents: "Array(3)",
      drinks: [
        "alcoholic",
        "others",
      ],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      names: "Array(2)",
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      site: "Overrided site name",
      tags: "Array(1)",
      title: "Page 4",
      url: "/pages/page4/",
    },
    src: {
      asset: undefined,
      ext: ".tmpl.ts",
      path: "/pages/2021-01-02-18-32_page4",
      remote: undefined,
      slug: "page4",
    },
  },
  {
    content: "<!DOCTYPE html>
<p>Content of Page 6</p>
",
    data: {
      children: "<p>Content of Page 6</p>
",
      colors: "Array(3)",
      content: "Content of Page 6
",
      date: [],
      documents: "Array(3)",
      drinks: [
        "alcoholic",
        "others",
      ],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      names: "Array(2)",
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      site: "Folder overrided site name",
      tags: "Array(2)",
      title: "Page 6",
      url: "/pages/page6/",
    },
    src: {
      asset: undefined,
      ext: ".md",
      path: "/pages/2021-12-29-page6",
      remote: undefined,
      slug: "page6",
    },
  },
  {
    content: "Content of Page 3",
    data: {
      colors: "Array(3)",
      content: "Content of Page 3",
      date: [],
      documents: "Array(3)",
      drinks: [
        "alcoholic",
        "others",
      ],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      names: "Array(2)",
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      site: "Folder overrided site name",
      tags: "Array(1)",
      title: "Page 3",
      url: "/page_3",
    },
    src: {
      asset: undefined,
      ext: ".tmpl.js",
      path: "/pages/page3",
      remote: undefined,
      slug: "page3",
    },
  },
  {
    content: "<!DOCTYPE html>
Content of Page 7",
    data: {
      children: "Content of Page 7",
      colors: "Array(3)",
      content: "Content of Page 7",
      date: [],
      documents: "Array(3)",
      drinks: [
        "alcoholic",
        "others",
      ],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      names: "Array(2)",
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      site: "Folder overrided site name",
      tags: "Array(3)",
      url: "/pages/subpage/page7/",
    },
    src: {
      asset: undefined,
      ext: ".tmpl.js",
      path: "/pages/subpage/page7",
      remote: undefined,
      slug: "page7",
    },
  },
]
`;
